odl/solvers/functional/functional.py

```diff
@@ -875,6 +876,149 @@ def convex_conj(self):
             self.linear_term)
 
 
+class FunctionalProduct(Functional, OperatorPointwiseProduct):
+
+    """Product ``p(x) = f(x) * g(x)`` of two functionals ``f`` and ``g``."""
+
+    def __init__(self, left, right):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        left, right : `Functional`
+            Functionals in the product. Need to have matching domains.
+
+        Examples
+        --------
+        Construct the functional || . ||_2^2 * 3
+
+        >>> import odl
+        >>> space = odl.rn(2)
+        >>> func1 = odl.solvers.L2NormSquared(space)
+        >>> func2 = odl.solvers.ConstantFunctional(space, 3)
+        >>> prod = odl.solvers.FunctionalProduct(func1, func2)
+        >>> prod([2, 3])  # expect (2**2 + 3**2) * 3 = 39
+        39.0
+        """
+        if not isinstance(left, Functional):
+            raise TypeError('`left` {} is not a `Functional` instance'
+                            ''.format(left))
+        if not isinstance(right, Functional):
+            raise TypeError('`right` {} is not a `Functional` instance'
+                            ''.format(right))
+
+        OperatorPointwiseProduct.__init__(self, left, right)
+        Functional.__init__(self, left.domain, linear=False,
+                            grad_lipschitz=np.nan)
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional.
+
+        Notes
+        -----
+        The derivative is computed using Leibniz's rule:
+
+        .. math::
+            [\\nabla (f g)](p) = g(p) [\\nabla f](p) + f(p) [\\nabla g](p)
+        """
+        func = self
+
+        class FunctionalProductGradient(Operator):
+
+            """Functional representing the derivative of ``f(.) * g(.)``."""
+
+            def _call(self, x):
+                return (func.right(x) * func.left.gradient(x) +
+                        func.left(x) * func.right.gradient(x))
+
+        return FunctionalProductGradient(self.domain, self.domain,
+                                         linear=True)
```

Comment:
I know I'm slow, and that this is already merged, but this isn't a linear operator, right?